src/pybind/mgr/mgr_util.py

```diff
@@ -23,6 +27,283 @@
+
+    def del_all_handles(self):
+        with self.lock:
+            for fs_name in list(self.connections.keys()):
+                logger.info("waiting for pending ops for '{}'".format(fs_name))
+                self._del_fs_handle(fs_name, wait=True)
+                logger.info("pending ops completed for '{}'".format(fs_name))
+            # no new connections should have been initialized since its
+            # guarded on shutdown.
+            assert len(self.connections) == 0
+
+
+class CephfsClient(object):
+    def __init__(self, mgr):
+        self.mgr = mgr
+        self.stopping = Event()
+        self.connection_pool = ConnectionPool(self.mgr)
+
+    def shutdown(self):
+        logger.info("shutting down")
+        # first, note that we're shutting down
+        self.stopping.set()
+        # third, delete all libcephfs handles from connection pool
+        self.connection_pool.del_all_handles()
+
+
+class GlobalLock(object):
+    """
+    Singleton design pattern taken from http://www.aleax.it/5ep.html
+
+    Global lock to serialize operations in mgr/volumes. This lock
+    is currently held when accessing (opening) a volume to perform
+    group/subvolume operations. Since this is a big lock, it's rather
+    inefficient -- but right now it's ok since mgr/volumes does not
+    expect concurrent operations via its APIs.
+
+    As and when features get added (such as clone, where mgr/volumes
+    would maintain subvolume states in the filesystem), there might
+    be a need to allow concurrent operations. In that case it would
+    be nice to implement an efficient path based locking mechanism.
+
+    See: https://people.eecs.berkeley.edu/~kubitron/courses/cs262a-F14/projects/reports/project6_report.pdf
+    """
+    _shared_state = {
+        'lock': Lock(),
+        'init': False
+    }
+
+    def __init__(self):
+        with self._shared_state['lock']:
+            if not self._shared_state['init']:
+                self._shared_state['init'] = True
+        # share this state among all instances
+        self.__dict__ = self._shared_state
+
+    @contextlib.contextmanager
+    def lock_op(self):
+        with self._shared_state['lock']:
+            yield
+
+
+@contextlib.contextmanager
+def open_filesystem(fsc, fs_name):
+    """
+    Open a file system for exclusive access.
+    This API is to be used as a context manager.
+
+    :param fsc: cephfs client instance
+    :param fs_name: fs name
+    :return: yields a fs handle (ceph filesystem handle)
+    """
+    if fsc.is_stopping():
+        raise CephfsConnectionException(-errno.ESHUTDOWN,
+                                        "shutdown in progress")
+
+    g_lock = GlobalLock()
+    fs_handle = fsc.connection_pool.get_fs_handle(fs_name)
+    try:
+        with g_lock.lock_op():
+            yield fs_handle
+    finally:
+        fsc.connection_pool.put_fs_handle(fs_name)
+
+
+@contextlib.contextmanager
+def open_filesystem_lockless(fsc, fs_name):
```

Comment:
I think we should just stick to providing `open_filesystem()` helper and not worry about lifting `GlobalLock` class here. Users can serialize fs access by wrapping `open_filesystem()` call within a lock.